File: Basic Data Structures/BST.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    struct Node *left;
    struct Node *right;
};

struct Node *insert(int val, struct Node *root){
    if (root == NULL){
        struct Node *temp = (struct Node*)malloc(sizeof(struct Node*));
        temp->data = val;
        temp->left = NULL;
        temp->right = NULL;
        return temp;
    }
    if (val < root->data){
        root->left = insert(val, root->left);
    } else {
        root->right = insert(val, root->right);
    }
    return root;
}

void inorder(struct Node *root){
    if (root == NULL)
        return;
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

struct Node* remove_node(int d, struct Node* root){
    if(root == NULL){
        return root;
    }

    if(d < root->data){
        root->left = remove_node(d, root->left);
        return root;
    } else if(d > root->data){
        root->right = remove_node(d, root->right);
        return root;
    } else if (root->data == d) {
        struct Node* temp = NULL;
        if(root->left == NULL && root->right == NULL){
            free(root);
            return temp;
        } else if (root->left == NULL) {
            temp = root->right;
            root->right = NULL;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            temp = root->left;
            root->left = NULL;
            free(root);
            return temp;
        } else {
            int min_in_rst = __INT_MAX__;
            temp = root->right;
            while(temp){
                min_in_rst = min_in_rst > temp->data ? temp->data : min_in_rst;
                temp = temp->left;
            }
            root->data = min_in_rst;
            root->right=remove_node(min_in_rst, root->right);
            return root;
        }
    } else return NULL;
}

int main(){
    int d;
    scanf("%d", &d);
    struct Node *root = NULL;
    while (d != -1){
        root = insert(d, root);
        scanf("%d", &d);
    }
    inorder(root);
    root = remove_node(3, root);
    printf("\n\n");
    inorder(root);
    return 0;
}

/*
10
5
1
8
20
30
15
6
9
7
-1
*/

File: Basic Data Structures/BST.py
<<<<<<< HEAD
# cook your dish here
class Node:
	data = -1
	left = None
	right = None

	def __init__(self, data):
		self.data = data

def insert(root, data):
	if(root == None):
		return Node(data)
	if data < root.data:
		root.left = insert(root.left, data)
	else:
		root.right = insert(root.right, data)
	return root

def preorder_print(root):
	if(root == None):
		return
	print(root.data, end = ", ")
	preorder_print(root.left)
	preorder_print(root.right)

def inorder_print(root):
	if(root == None):
		return
	inorder_print(root.left)
	print(root.data, end = ", ")
	inorder_print(root.right)

def postorder_print(root):
	if(root == None):
		return
	postorder_print(root.left)
	postorder_print(root.right)
	print(root.data, end = ", ")

def levelorder_levelwise(root):
	q = []
	q.append(root)
	q.append(None)
	while q:
		curr = q.pop(0)
		if not q:
			break
		if curr == None:
			print()
			q.append(None)
			continue
		print(curr.data, end = ", ")
		if curr.left:
			q.append(curr.left)
		if curr.right:
			q.append(curr.right)

def display(li):
	for ele in li:
		print(ele.data, end = ", ")
	print()
	print("***********")

def max_depth_nodes(root):
	nodes = []
	q = []
	q.append(root)
	nodes.append(root)
	q.append(None)
	while len(q):
		cur = q.pop(0)
		if len(q) == 0:
		    break
		if cur == None:
		    q.append(None)
		    continue
		nodes.append(cur)
		if cur.left:
			q.append(cur.left)
		if cur.right:
			q.append(cur.right)
	return nodes

li = [int(x) for x in input().split()]
root = None
for ele in li:
	root = insert(root, ele)
print("Pre-order: ")
preorder_print(root)
print()

print("In-order: ")
inorder_print(root)
print()

print("Post-order: ")
postorder_print(root)
print()

print("levelorder levelwise: ")
levelorder_levelwise(root)
print()

print("****************")
mdn = max_depth_nodes(root)
for ele in mdn:
	print(ele.data, end = ", ")
=======
# cook your dish here
class Node:
	data = -1
	left = None
	right = None

	def __init__(self, data):
		self.data = data

def insert(root, data):
	if(root == None):
		return Node(data)
	if data < root.data:
		root.left = insert(root.left, data)
	else:
		root.right = insert(root.right, data)
	return root

def preorder_print(root):
	if(root == None):
		return
	print(root.data, end = ", ")
	preorder_print(root.left)
	preorder_print(root.right)

def inorder_print(root):
	if(root == None):
		return
	inorder_print(root.left)
	print(root.data, end = ", ")
	inorder_print(root.right)

def postorder_print(root):
	if(root == None):
		return
	postorder_print(root.left)
	postorder_print(root.right)
	print(root.data, end = ", ")

def levelorder_levelwise(root):
	q = []
	q.append(root)
	q.append(None)
	while q:
		curr = q.pop(0)
		if not q:
			break
		if curr == None:
			print()
			q.append(None)
			continue
		print(curr.data, end = ", ")
		if curr.left:
			q.append(curr.left)
		if curr.right:
			q.append(curr.right)

def display(li):
	for ele in li:
		print(ele.data, end = ", ")
	print()
	print("***********")

def max_depth_nodes(root):
	nodes = []
	q = []
	q.append(root)
	nodes.append(root)
	q.append(None)
	while len(q):
		cur = q.pop(0)
		if len(q) == 0:
		    break
		if cur == None:
		    q.append(None)
		    continue
		nodes.append(cur)
		if cur.left:
			q.append(cur.left)
		if cur.right:
			q.append(cur.right)
	return nodes

li = [int(x) for x in input().split()]
root = None
for ele in li:
	root = insert(root, ele)
print("Pre-order: ")
preorder_print(root)
print()

print("In-order: ")
inorder_print(root)
print()

print("Post-order: ")
postorder_print(root)
print()

print("levelorder levelwise: ")
levelorder_levelwise(root)
print()

print("****************")
mdn = max_depth_nodes(root)
for ele in mdn:
	print(ele.data, end = ", ")
>>>>>>> 24a80fac1064910ae51f09e81777487a1aaee3b1
print()